### ISSUE TITLE
ERR_HTTP2_PROTOCOL_ERROR when zstd module enabled

### ISSUE THREAD

**Issue by Arien02** [tokers/zstd-nginx-module#32]
Hi all!

I'm facing this error with zstd module enabled and zstd browser:

ERR_HTTP2_PROTOCOL_ERROR

Web page it's not loaded at all, not any single png, or text, the response it's just empty.

![Error_1](https://github.com/tokers/zstd-nginx-module/assets/117276287/137ccb77-e5ef-4f1b-adde-0d2f026973cb)

Tried with different parameters and mime types, without effects, activated nginx debug without any significant info.

Can anybody point me in any direction to solve this?

**NGINX info:**
nginx version: nginx/1.25.4 (https://deb.myguard.nl/nginx-modules/)
built with OpenSSL 3.0.13+quic 30 Jan 2024
TLS SNI support enabled
configure arguments: --conf-path=/etc/nginx/nginx.conf --error-log-path=stderr --http-client-body-temp-path=/var/lib/nginx/body --http-fastcgi-temp-path=/var/lib/nginx/fastcgi --http-log-path=/var/log/nginx/access.log --http-proxy-temp-path=/var/lib/nginx/proxy --http-scgi-temp-path=/var/lib/nginx/scgi --http-uwsgi-temp-path=/var/lib/nginx/uwsgi --lock-path=/var/lock/nginx.lock --modules-path=/usr/lib/nginx/modules --pid-path=/run/nginx.pid --prefix=/usr/share/nginx --with-compat --with-file-aio --with-threads --with-cc-opt='-g -ffile-prefix-map=/build/nginx-1.25.4=. -fstack-protector-strong -Wformat -Werror=format-security -Ofast -flto -DNGX_HTTP_HEADERS -DNGX_ZLIB_NG=1 -w -DTCP_FASTOPEN=23 -fomit-frame-pointer -pipe -fPIC -Wdate-time -D_FORTIFY_SOURCE=2' --with-ld-opt='-Wl,-z,relro -Wl,-z,now -fPIC' --with-pcre-jit --with-http_dav_module --with-http_ssl_module --with-http_v2_module --with-http_v3_module --with-stream=dynamic --with-stream_realip_module --with-stream_ssl_module --with-stream_ssl_preread_module --build=https://deb.myguard.nl/nginx-modules/ --with-http_addition_module --with-http_auth_request_module --with-http_degradation_module --with-http_flv_module --with-http_geoip_module=dynamic --with-http_gunzip_module --with-http_gzip_static_module --with-http_image_filter_module=dynamic --with-http_mp4_module --with-http_perl_module=dynamic --with-http_random_index_module --with-http_realip_module --with-http_secure_link_module --with-http_slice_module --with-http_stub_status_module --with-http_sub_module --with-http_xslt_module=dynamic --with-mail=dynamic --with-mail_ssl_module --with-stream_geoip_module=dynamic

**NGINX module loaded:**
load_module modules/ngx_http_zstd_filter_module.so;
load_module modules/ngx_http_zstd_static_module.so;

**ZSTD module parameters:**
zstd on;
zstd_min_length 256;
zstd_comp_level 3;
zstd_buffers 32 8k;
zstd_types application/atom+xml application/geo+json application/javascript application/x-javascript application/json application/ld+json application/manifest+json application/rdf+xml application/rss+xml application/vnd.ms-fontobject application/wasm application/x-web-app-manifest+json application/xhtml+xml application/xml application/xml+rss font/eot font/otf font/ttf image/bmp image/svg+xml image/vnd.microsoft.icon image/x-icon text/cache-manifest text/calendar text/css text/javascript text/markdown text/plain text/xml text/vcard text/vnd.rim.location.xloc text/vtt text/x-component text/x-cross-domain-policy;

**NGINX debug:**
2024/04/06 09:37:31 [alert] 55733#55733: *56 the http output chain is empty while sending to client, client: xx.xxx.xxx.xx, server: proseg.es, request: "GET / HTTP/2.0", upstream: "fastcgi://unix:/var/lib/php8.2-fpm/web13.sock:", host: "www.proseg.es"

Thank you!

**Comment by tokers** [comment]
@Arien02 Could you provide the Nginx error log and the response header?

**Comment by Arien02** [comment]
@tokers Hi there!
The response header is empty, no response at all:
![zstd_error](https://github.com/tokers/zstd-nginx-module/assets/117276287/17daebcf-342d-4ad0-b04d-b5022bc66dfd)

This is curl response with zstd module enabled (curl -v --compressed https://www.proseg.es):

`* using HTTP/2
* h2h3 [:method: GET]
* h2h3 [:path: /]
* h2h3 [:scheme: https]
* h2h3 [:authority: www.proseg.es]
* h2h3 [user-agent: curl/7.88.1]
* h2h3 [accept: */*]
* h2h3 [accept-encoding: deflate, gzip, br, zstd]
* Using Stream ID: 1 (easy handle 0x5bc6feebdc80)
> GET / HTTP/2
> Host: www.proseg.es
> user-agent: curl/7.88.1
> accept: */*
> accept-encoding: deflate, gzip, br, zstd
>
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* old SSL session ID is stale, removing
* HTTP/2 stream 1 was not closed cleanly: INTERNAL_ERROR (err 2)
* Connection #0 to host www.proseg.es left intact
curl: (92) HTTP/2 stream 1 was not closed cleanly: INTERNAL_ERROR (err 2)`

And this is without zstd module enabled:

`* using HTTP/2
* h2h3 [:method: GET]
* h2h3 [:path: /]
* h2h3 [:scheme: https]
* h2h3 [:authority: www.proseg.es]
* h2h3 [user-agent: curl/7.88.1]
* h2h3 [accept: */*]
* h2h3 [accept-encoding: deflate, gzip, br, zstd]
* Using Stream ID: 1 (easy handle 0x62f3f6098c80)
> GET / HTTP/2
> Host: www.proseg.es
> user-agent: curl/7.88.1
> accept: */*
> accept-encoding: deflate, gzip, br, zstd
>
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* old SSL session ID is stale, removing
< HTTP/2 200
`

This is the error.log relative with the issue:

`2024/04/30 19:05:59 [alert] 26283#26283: *39340 the http output chain is empty while sending to client, client: xx.xxx.xx.xxx, server: proseg.es, request: "GET / HTTP/2.0", upstream: "fastcgi://unix:/var/lib/php8.2-fpm/web13.sock:", host: "www.proseg.es"`

Thank you for your time.

**Comment by Arien02** [comment]
@tokers Hi again!

I have just compiled again zstd module with your last week commit, and suddlenly it is working fine right now.
So it seems to be related with this closed issue: https://github.com/tokers/zstd-nginx-module/issues/30

![zstd_error2](https://github.com/tokers/zstd-nginx-module/assets/117276287/5c4f3f97-0229-4ed5-b3e2-633cf3d7a7cc)

I will reopen this one if I found any other related problem.

Thank you again for your work!
